cpp/src/arrow/compute/kernels/scalar_temporal_test.cc

```diff
@@ -0,0 +1,107 @@
+  auto iso_calendar = ArrayFromJSON(iso_calendar_type,
+                                    R"([{"iso_year": 1970, "iso_week": 1, "weekday": 4},
+                        {"iso_year": 2000, "iso_week": 9, "weekday": 2},
+                        {"iso_year": 1899, "iso_week": 52, "weekday": 7},
+                        {"iso_year": 2033, "iso_week": 20, "weekday": 3}, null])");
+  auto quarter = "[1, 1, 1, 2, null]";
+  auto hour = "[0, 23, 0, 3, null]";
+  auto minute = "[0, 23, 59, 33, null]";
+  auto second = "[59.123456789, 23.999999999, 20.001001001, 20.0, null]";
+  auto millisecond = "[123, 999, 1, 0, null]";
+  auto microsecond = "[456, 999, 1, 0, null]";
+  auto nanosecond = "[789, 999, 1, 0, null]";
+  auto subsecond = "[123456789, 999999999, 1001001, 0, null]";
+
+  CheckScalarUnary("year", unit, times, int64(), year);
+  CheckScalarUnary("month", unit, times, int64(), month);
+  CheckScalarUnary("day", unit, times, int64(), day);
+  CheckScalarUnary("day_of_week", unit, times, int64(), day_of_week);
+  CheckScalarUnary("day_of_year", unit, times, int64(), day_of_year);
+  CheckScalarUnary("iso_year", unit, times, int64(), iso_year);
+  CheckScalarUnary("iso_week", unit, times, int64(), iso_week);
+  CheckScalarUnary("iso_calendar", timestamps, iso_calendar);
+  CheckScalarUnary("quarter", unit, times, int64(), quarter);
+  CheckScalarUnary("hour", unit, times, int64(), hour);
+  CheckScalarUnary("minute", unit, times, int64(), minute);
+  CheckScalarUnary("second", unit, times, float64(), second);
+  CheckScalarUnary("millisecond", unit, times, int64(), millisecond);
+  CheckScalarUnary("microsecond", unit, times, int64(), microsecond);
+  CheckScalarUnary("nanosecond", unit, times, int64(), nanosecond);
+  CheckScalarUnary("subsecond", unit, times, int64(), subsecond);
+}
+
+TEST(ScalarTemporalTest, TestZonedTemporalComponentExtraction) {
+  std::string timezone = "Etc/UTC-2";
+  const char* times =
+      R"(["1970-01-01T00:00:59.123456789","2000-02-29T23:23:23.999999999",
+          "1899-01-01T00:59:20.001001001","2033-05-18T03:33:20.000000000", null])";
+  auto unit = timestamp(TimeUnit::NANO, timezone);
+  auto timestamps = ArrayFromJSON(unit, times);
+
+  ASSERT_RAISES(Invalid, Year(timestamps));
+  ASSERT_RAISES(Invalid, Month(timestamps));
+  ASSERT_RAISES(Invalid, Day(timestamps));
+  ASSERT_RAISES(Invalid, DayOfWeek(timestamps));
+  ASSERT_RAISES(Invalid, DayOfYear(timestamps));
+  ASSERT_RAISES(Invalid, ISOYear(timestamps));
+  ASSERT_RAISES(Invalid, ISOWeek(timestamps));
+  ASSERT_RAISES(Invalid, ISOCalendar(timestamps));
+  ASSERT_RAISES(Invalid, Quarter(timestamps));
+  ASSERT_RAISES(Invalid, Hour(timestamps));
+  ASSERT_RAISES(Invalid, Minute(timestamps));
+  ASSERT_RAISES(Invalid, Second(timestamps));
+  ASSERT_RAISES(Invalid, Millisecond(timestamps));
+  ASSERT_RAISES(Invalid, Microsecond(timestamps));
+  ASSERT_RAISES(Invalid, Nanosecond(timestamps));
+  ASSERT_RAISES(Invalid, Subsecond(timestamps));
```

Comment:
Wouldn't ignoring be a bit unexpected? It would be relatively easy to add timezone processing now - [see here](https://github.com/apache/arrow/commit/36abbcb512b2b45c8b5615fe29074be2afbd01d8).

Reply:
What is the expected semantics? It seems to me that the results would simply returned in the local timezone, i.e. if I have a timestamp for '2000-01-01 00:00:00' in the UTC+1 timezone, then the year is 2000... in the UTC+1 timezone.

If users want to convert to UTC first, perhaps we want a separate kernel for that, but I see no reason to do that conversion in the datetime component extraction kernels.

For example in Python:
```python
>>> t = dateutil.tz.gettz('Europe/Paris')
>>> dt = datetime.datetime(2000,1,1,0,0,0, tzinfo=t)
>>> dt
datetime.datetime(2000, 1, 1, 0, 0, tzinfo=tzfile('/usr/share/zoneinfo/Europe/Paris'))
>>> dt.year
2000
>>> t.utcoffset(dt)
datetime.timedelta(seconds=3600)
>>> dt.utctimetuple().tm_year
1999
```

Reply:
Arrow timestamp is UTC unix time with optional timezone information (if I understand correctly).
So if I want to know e.g. hour in my timezone I need to convert to my timezone at some point.
Timezone conversion kernel is an option, but then timestamp becomes local time instead of unix time?

Reply:
Ah... It seems you're right indeed. 
```
          /// * If the time zone is set to a valid value, values can be displayed as
          ///   "localized" to that time zone, even though the underlying 64-bit
          ///   integers are identical to the same data stored in UTC. Converting
          ///   between time zones is a metadata-only operation and does not change the
          ///   underlying values
```

I will ask on the ML for clarification.

Reply:
In any case, this is mostly a documentation issue. If the timestamp value is always in UTC, then "year" and friends will return the year expressed in the UTC timezone. I don't think there is any reason to forbid the operation, just document it properly.

Reply:
ML discussion on this would be good indeed.
About raising when timezone is present - as a user I would be slightly surprised but could live with it.
@joris [previously suggested raising](https://github.com/apache/arrow/pull/10176#issuecomment-828254868).

I'm indifferent on this but would like add timezone processing ([like so](https://github.com/rok/arrow/commit/36abbcb512b2b45c8b5615fe29074be2afbd01d8)).

Reply:
Indeed, the storage model of timezone-aware data is different compared to stdlib python, we are always storing the UTC timestamp (and not the localized timestamp).

> In any case, this is mostly a documentation issue. If the timestamp value is always in UTC, then "year" and friends will return the year expressed in the UTC timezone. I don't think there is any reason to forbid the operation, just document it properly.

I personally disagree with this. If you have a timezone-aware timestamp, as a user I would expect to get the result in *local*time. If I want to have the UTC version, I can easily first convert to UTC (which is a no-op). 
For "year" it most of the time doesn't change much (only the hours around New Year), but if I have a timestamp "2021-06-01 09:00:00+02:00" (CEST), and I ask for "hour", I expect to get 9 and not 7.

Reply:
Hmm, that makes sense.

Reply:
Let's leave it as is then. I'll open a followup JIRA to add timezone processing.